### DISCUSSION TITLE
Shared volume mount

### DISCUSSION THREAD

**Discussion by mflynn-lanl** [microbiomedata/nmdc-edge#207] [General]
I see different files when I list the directories from the web app than when I list them from the MC login node for the shared volume 

```console
[user@login1 ~]$ ll /project/io/upload/files/
total 1183288
drwxrwxrwx. 2 nobody nobody       165 Jun 17 17:00 .
drwxrwxrwx. 4 nobody nobody        42 Jun  5 23:18 ..
-rw-r--r--. 1 nobody nobody 385094750 Jun 11 20:29 EMLcJxh9BENoBvJo.fastq.gz
-rw-r--r--. 1 nobody nobody 206954680 Jun 13 16:45 jhOseEM0hhmxe7Sf.fastq.gz
-rw-r--r--. 1 nobody nobody 206954680 Jun 14 20:09 NeuQx452YkZO85JV.fastq.gz
-rw-r--r--. 1 nobody nobody 206333469 Jun 17 16:44 poXDwwXxRrovLwvJ.fastq
-rw-r--r--. 1 nobody nobody 206333469 Jun 17 17:00 Ud744ifFzhvH54GP.fastq
```

```console
exouser@nmdc-edge-web-app-via-caddy-20240407:~$ ll /mnt/io/upload/files/
total 201508
drwxrwx--- 2 webuser cromwell      4096 Jun 20 21:42 ./
drwxrwx--- 4 webuser cromwell      4096 Jun 20 07:57 ../
-rw-r--r-- 1 webuser cromwell 206333469 Jun 20 20:25 1KvhBDCrRR0vsQxo.fastq
-rw-rw-r-- 1 exouser exouser          0 Jun 20 21:42 test.txt
```

All of the other directories have different files as well

**Comment by mflynn-lanl** [comment]
I tried creating a file directly on the web app file system in `/mnt/io/upload/files`, but it didn't show up in `/project/io/upload/files`. I tried the reverse but that file didn't appear in `/mnt/io/upload/files`

**Reply by edwins** [reply]
that sounds unusual, unless there's some sort of caching. from the webapp side of things, can you confirm that the nfs share is properly mounted?

**Reply by mflynn-lanl** [reply]
It seems like it was working before, there are files that were added at one point but now there is the disconnect. What can I do to confirm that the nfs share is mounted properly?

**Reply by mflynn-lanl** [reply]
I wonder if all of the permissions/ownership changes caused problems

**Reply by edwins** [reply]
Hi @mflynn-lanl ,

If your webapp vm is a new vm, you will need to update the relevant  magic castle security group to give access to the new vm's ip. 

You can use the `mount` command to see which mounts are active. It should show the magic castle project share.

Let me know if you're still having issue's connecting the project nfs share from magic castle.

**Reply by mflynn-lanl** [reply]
I don't see `/mnt` in the output from `mount`. 
I tried running the mount command I ran before 
`exouser@nmdc-edge-web-app-via-caddy-20240407:~$ mount -t nfs4 -o proto=tcp,nosuid,nolock,noatime,actimeo=3,nfsvers=4.2,seclabel,x-systemd.automount,x-systemd.mount-timeout=30,_netdev 10.0.61.212:/project /mnt/
mount.nfs4: failed to apply fstab options`

**Reply by edwins** [reply]
`mount` is typically requires root privileges. See if doing `sudo mount` solves that specific error.

**Reply by mflynn-lanl** [reply]
That worked. Now I see the same files on the shared drive from both the webapp VM and the MC login node. However, I'm still getting an error: `mkdir: cannot create directory '/project/io': Read-only file system 
cp: cannot create regular file /project/io/projects/fPyybyo3BB2dLKyF/output/ReadsQC/': No such file or directory`
even though that directory was created when the project was submitted by the webapp.

**Reply by mflynn-lanl** [reply]
I get this error when I submit the workflow(curl command) from the webapp or from the MC login node. 
permissions from the webapp VM:

```console
exouser@nmdc-edge-web-app-via-caddy-20240407:/mnt/io$ ll projects/fPyybyo3BB2dLKyF/
total 36
drwxrwxrwx 4 webuser cromwell   179 Jun 28 17:00 ./
drwxrwxrwx 3 webuser cromwell    30 Jun 25 17:53 ../
-rwxrwxrwx 1 webuser cromwell   272 Jun 25 17:53 conf.json*
-rwxrwxrwx 1 webuser cromwell 16449 Jun 25 17:57 cromwell_job_metadata.json*
drwxrwxrwx 2 webuser cromwell    26 Jun 25 17:54 input/
-rwxrwxrwx 1 webuser cromwell    85 Jun 25 17:57 log.txt*
drwxrwxrwx 3 webuser cromwell    21 Jun 25 17:54 output/
-rwxrwxrwx 1 webuser cromwell  1530 Jun 25 17:54 pipeline.wdl*
-rwxrwxrwx 1 webuser cromwell   382 Jun 25 17:54 pipeline_inputs.json*
```

From the MC login node:

```console
[user@login1 ~]$ ll /project/io/projects/fPyybyo3BB2dLKyF/
total 36
drwxrwxrwx. 4 user user   179 Jun 28 17:00 .
drwxrwxrwx. 3 user user    30 Jun 25 17:53 ..
-rwxrwxrwx. 1 user user   272 Jun 25 17:53 conf.json
-rwxrwxrwx. 1 user user 16449 Jun 25 17:57 cromwell_job_metadata.json
drwxrwxrwx. 2 user user    26 Jun 25 17:54 input
-rwxrwxrwx. 1 user user    85 Jun 25 17:57 log.txt
drwxrwxrwx. 3 user user    21 Jun 25 17:54 output
-rwxrwxrwx. 1 user user   382 Jun 25 17:54 pipeline_inputs.json
-rwxrwxrwx. 1 user user  1530 Jun 25 17:54 pipeline.wdl
```

**Reply by edwins** [reply]
I don't see a screenshot of the error. Can you repaste?

**Reply by mflynn-lanl** [reply]
I figured out the problem. It was a problem with my Cromwell config file. The directory `/project/io` wasn't being mounted for the Singularity exec command

**Reply by mflynn-lanl** [reply]
I am able to submit workflows from the website! Everything is good, at least for now

**Reply by edwins** [reply]
I'm glad to hear. I'm looking forward to see when everything is connected completely, and to learn about all the customization that was required to adapt to magic castle.

**Comment by mflynn-lanl** [comment]
@edwins I'm still having trouble with file ownership on the MC cluster. on the login node some of the files are owned by user `mcflynn`: 
```
[user@login1 projects]$ ll ../nmdc-edge/data/workflow/templates/
total 148
drwxrwxrwx. 2 user    user     4096 Sep 27 18:20 .
drwxrwxrwx. 4 user    user       34 Sep  6 01:47 ..
-rw-r--r--. 1 mcflynn mcflynn   969 Sep 19 18:00 enviroMS_inputs.tmpl
-rw-r--r--. 1 mcflynn mcflynn  1531 Sep 19 18:00 enviroMS_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn  1538 Sep 19 18:00 enviroMS_wdl.tmpl.backup
-rw-r--r--. 1 mcflynn mcflynn   374 Sep 19 18:00 geNomad_inputs.tmpl
-rw-r--r--. 1 mcflynn mcflynn   543 Sep 19 18:00 geNomad_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn   163 Sep 19 18:00 metaAnnotation_inputs.tmpl
-rw-r--r--. 1 mcflynn mcflynn   104 Sep 19 18:00 metaAnnotation_options.json
-rw-r--r--. 1 mcflynn mcflynn   867 Sep 19 18:00 metaAnnotation_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn   327 Sep 19 18:00 metaAssembly_inputs.tmpl
-rw-r--r--. 1 mcflynn mcflynn  1585 Sep 19 18:00 metaAssembly_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn    84 Sep 19 18:00 metaG_options.tmpl
-rw-r-----. 1 user    user      907 Sep 19 22:13 metaMAGs_inputs.tmpl
-rw-r-----. 1 user    user     2485 Sep 19 22:13 metaMAGs_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn   685 Sep 19 18:00 metaProteomics_inputs.tmpl
-rw-r-----. 1 user    user      291 Sep 19 22:13 metaT_inputs.tmpl
-rw-r-----. 1 user    user      494 Sep 19 22:13 metaT_wdl.tmpl
-rwxrwxrwx. 1 user    user     1314 Sep  6 04:15 metagenome_pipeline_inputs.tmpl
-rw-r--r--. 1 mcflynn mcflynn   105 Sep 19 18:00 metagenome_pipeline_options.json
-rw-r-----. 1 user    user    21608 Sep 19 22:13 metagenome_pipeline_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn 15980 Sep 19 18:00 metagenome_pipeline_wdl.tmpl.bak
-rw-r--r--. 1 mcflynn mcflynn   255 Sep 19 18:00 readbasedAnalysis_inputs.tmpl
-rw-r--r--. 1 mcflynn mcflynn  1641 Sep 19 18:00 readbasedAnalysis_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn   327 Sep 19 18:00 readsQC_inputs.tmpl
-rw-r-----. 1 user    user     1491 Sep 19 22:13 readsQC_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn    95 Sep 19 18:00 sra2fastq_inputs.tmpl
-rw-r--r--. 1 mcflynn mcflynn   301 Sep 19 18:00 sra2fastq_wdl.tmpl
-rw-r--r--. 1 mcflynn mcflynn   930 Sep 19 18:00 virus_plasmid_inputs.tmpl
-rw-r--r--. 1 mcflynn mcflynn  1336 Sep 19 18:00 virus_plasmid_wdl.tmpl
```
When I look at them from the web app VM I see that some of the files are owned by `exouser`: 
```
(base) exouser@nmdc-test:~$ ll /mnt/io/nmdc-edge/data/workflow/templates/
total 148
drwxrwxrwx 2 webuser webuser  4096 Sep 27 18:20 ./
drwxrwxrwx 4 webuser webuser    34 Sep  6 01:47 ../
-rw-r--r-- 1 exouser exouser   969 Sep 19 18:00 enviroMS_inputs.tmpl
-rw-r--r-- 1 exouser exouser  1531 Sep 19 18:00 enviroMS_wdl.tmpl
-rw-r--r-- 1 exouser exouser  1538 Sep 19 18:00 enviroMS_wdl.tmpl.backup
-rw-r--r-- 1 exouser exouser   374 Sep 19 18:00 geNomad_inputs.tmpl
-rw-r--r-- 1 exouser exouser   543 Sep 19 18:00 geNomad_wdl.tmpl
-rw-r--r-- 1 exouser exouser   163 Sep 19 18:00 metaAnnotation_inputs.tmpl
-rw-r--r-- 1 exouser exouser   104 Sep 19 18:00 metaAnnotation_options.json
-rw-r--r-- 1 exouser exouser   867 Sep 19 18:00 metaAnnotation_wdl.tmpl
-rw-r--r-- 1 exouser exouser   327 Sep 19 18:00 metaAssembly_inputs.tmpl
-rw-r--r-- 1 exouser exouser  1585 Sep 19 18:00 metaAssembly_wdl.tmpl
-rw-r--r-- 1 exouser exouser    84 Sep 19 18:00 metaG_options.tmpl
-rw-r----- 1 webuser webuser   907 Sep 19 22:13 metaMAGs_inputs.tmpl
-rw-r----- 1 webuser webuser  2485 Sep 19 22:13 metaMAGs_wdl.tmpl
-rw-r--r-- 1 exouser exouser   685 Sep 19 18:00 metaProteomics_inputs.tmpl
-rw-r----- 1 webuser webuser   291 Sep 19 22:13 metaT_inputs.tmpl
-rw-r----- 1 webuser webuser   494 Sep 19 22:13 metaT_wdl.tmpl
-rwxrwxrwx 1 webuser webuser  1314 Sep  6 04:15 metagenome_pipeline_inputs.tmpl*
-rw-r--r-- 1 exouser exouser   105 Sep 19 18:00 metagenome_pipeline_options.json
-rw-r----- 1 webuser webuser 21608 Sep 19 22:13 metagenome_pipeline_wdl.tmpl
-rw-r--r-- 1 exouser exouser 15980 Sep 19 18:00 metagenome_pipeline_wdl.tmpl.bak
-rw-r--r-- 1 exouser exouser   255 Sep 19 18:00 readbasedAnalysis_inputs.tmpl
-rw-r--r-- 1 exouser exouser  1641 Sep 19 18:00 readbasedAnalysis_wdl.tmpl
-rw-r--r-- 1 exouser exouser   327 Sep 19 18:00 readsQC_inputs.tmpl
-rw-r----- 1 webuser webuser  1491 Sep 19 22:13 readsQC_wdl.tmpl
-rw-r--r-- 1 exouser exouser    95 Sep 19 18:00 sra2fastq_inputs.tmpl
-rw-r--r-- 1 exouser exouser   301 Sep 19 18:00 sra2fastq_wdl.tmpl
-rw-r--r-- 1 exouser exouser   930 Sep 19 18:00 virus_plasmid_inputs.tmpl
-rw-r--r-- 1 exouser exouser  1336 Sep 19 18:00 virus_plasmid_wdl.tmpl
```
I am unable to change the ownership from exouser even with `sudo` from the web app VM or as user on the login node.

**Reply by edwins** [reply]
Try using uid rather than username. that might work better.

**Reply by mflynn-lanl** [reply]
@edwins I don't remember if using the uid worked before, but I created a new cluster and it doesn't work now.  I am not able to chmod or chown the /project directory, either as exouser or webuser (uid 60006) on the web app vm nor as mcflynn or user on the login node of the cluster.

**Comment by mflynn-lanl** [comment]
I'm also having problem with Cromwell not finding files:
```
/project/io/nmdc-edge/data/project/aaron/data/3603_017-B2_7-306-5-354__GAGGACAG-GCCGAATC__S1_R1_001.fastq.gz doesn't exist
	File not found /project/io/nmdc-edge/data/project/aaron/data/3603_017-B2_7-306-5-354__GAGGACAG-GCCGAATC__S1_R1_001.fastq.gz
```

```
(base) [user@login1 ~]$ ll /project/io/nmdc-edge/data/project/aaron/data/3603_017-B2_7-306-5-354__GAGGACAG-GCCGAATC__S1_R1_001.fastq.gz
-rwxrwxrwx. 1 user user 13071860085 Sep 30 18:55 /project/io/nmdc-edge/data/project/aaron/data/3603_017-B2_7-306-5-354__GAGGACAG-GCCGAATC__S1_R1_001.fastq.gz
```

**Comment by mflynn-lanl** [comment]
The weird thing is that I tried to run the exact same thing I ran before that failed due to lack of disk space. I deleted some data and when I tried to re-submit it but it failed the second time. I get the same thing for other files as well

**Reply by edwins** [reply]
Just to confirm, cromwell is installed on the same head node for the cluster that you're submitting to? And, can you see the file on the head node if you look at it through the filesystem on the head node? if true, i'm unsure why cromwell cannot see the file if you can perform an `ls` on the same login node

**Comment by mflynn-lanl** [comment]
Yes, that is all correct. Initially, Cromwell could see the files. I have had some problems on the other cluster, the one I am using for NMDC-EDGE, but there I saw changes in permissions and/or file ownership. With this cluster, the file permissions are 777 and somehow things have changed from when I initially submitted Cromwell workflows. Everything is exactly the same, the file locations, permissions, ownership and the Cromwell server, except now I get the file not found errors.
________________________________
From: Edwin Skidmore ***@***.***>
Sent: Friday, October 11, 2024 11:08 AM
To: microbiomedata/nmdc-edge ***@***.***>
Cc: Flynn, Mark C ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [microbiomedata/nmdc-edge] Shared volume mount (Discussion #207)


Just to confirm, cromwell is installed on the same head node for the cluster that you're submitting to? And, can you see the file on the head node if you look at it through the filesystem on the head node? if true, i'm unsure why cromwell cannot see the file if you can perform an ls on the same login node
